### DISCUSSION TITLE
Cascade Delete Trigger Hooks

### DISCUSSION THREAD

**Discussion by LeCyreaxYT** [pocketbase/pocketbase#6087] [Q&A] [answered]
Hi Friends,

I have a problem with the hooks in my code. I'm using the JS hooks option, and everything works fine, but there's an issue with my hooks when I want to trigger a "Webhook."

To clarify, here's an example:

Users
Sanctions -> (relation: user, ...)
When I delete a user, it triggers a webhook with some data. However, the problem is that a cascade delete happens on Sanctions, which also triggers the sanctions webhook to Discord and causes it to crash.

I don't want the cascade delete to trigger the sanctions webhook. Is there a way to prevent this trigger or detect whether it's a cascade delete? Is there a solution in JS, or do I need to switch to Go?

Example Code:
```js
onRecordDelete((e) => {
  const utils = require(`${__hooks}/utils.js`);
  const webhook = utils.getWebhookSanctionUrl(e.app.settings().meta.appURL);
  utils.logSanction(e, 'delete', webhook);

  e.next();
}, 'sanction');

onRecordAfterUpdateSuccess((e) => {
  const utils = require(`${__hooks}/utils.js`);
  const webhook = utils.getWebhookSanctionUrl(e.app.settings().meta.appURL);
  utils.logSanction(e, 'update', webhook);

  e.next();
}, 'sanction');

onRecordAfterCreateSuccess((e) => {
  const utils = require(`${__hooks}/utils.js`);
  const webhook = utils.getWebhookSanctionUrl(e.app.settings().meta.appURL);
  utils.logSanction(e, 'create', webhook);

  e.next();
}, 'sanction');
```

**Comment by ganigeorgiev** [comment]
If you don't want cascade delete then you can disable it from the `relation` field options but there is no knob that will enable the cascade delete conditionally.

Depending on your use case an alternative could be to attach to the `onRecordDeleteRequest` hook instead because in this case the `sanction` delete hook handler will be invoked only on a direct DELETE API request to the `sanction` collection.

**Reply by LeCyreaxYT** [reply]
Thanks for the fast response. The problem I have with the `onRecordDeleteRequest` is that it triggers a webhook, but the request fails due to API rules. I want the webhook to be triggered only if the record is 100% deleted. There is no other way to do this?

**Reply by ganigeorgiev** [reply, marked as answer]
If I understand what you want then triggering your hook AFTER the `e.next()` call should suffice, aka.:

```js
onRecordDeleteRequest((e) => {
  e.next(); // if the delete fails this throws

  const utils = require(`${__hooks}/utils.js`);
  const webhook = utils.getWebhookSanctionUrl(e.app.settings().meta.appURL);
  utils.logSanction(e, 'delete', webhook);
}, 'sanction');
```

**Reply by ganigeorgiev** [reply]
The only case where I could think the above may fail is if you run the delete as part of the [batch web API](https://pocketbase.io/docs/api-records/#batch-createupdateupsertdelete-records) and some of the other batch requests fails because in this case the entire batch transaction will be rollbacked.

**Reply by ganigeorgiev** [reply]
Another slightly hacky approach if you want the above to work even with the batch web API could be to set a dummy record field from the request hook and then check it in the record db hook to determine whether the delete operation was from the request or from somewhere else (in your case most likely by a cascade delete operation). For example:

```js
onRecordDeleteRequest((e) => {
  e.record.set("isDeleteFromRequest", true)
  e.next();
}, 'sanction');

onRecordAfterDeleteSuccess((e) => {
  if (e.record.get("isDeleteFromRequest")) {
    const utils = require(`${__hooks}/utils.js`);
    const webhook = utils.getWebhookSanctionUrl(e.app.settings().meta.appURL);
    utils.logSanction(e, 'delete', webhook);
  }

  e.next();
}, 'sanction');
```

But again, there is no knob that will apply the cascade delete conditionally or option to detect the origin of the delete operation. Technically it is possible to try to add meta information what has triggered the hook but without the full picture of your use case and why what you want is even needed in the first place, I'm not willing to invest time on this.

**Reply by LeCyreaxYT** [reply]
Thanks for these tips. It's working now.
